repository/lock_cleaner.go

```diff
@@ -0,0 +1,54 @@
+package repository
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ttab/elephantine"
+	"github.com/ttab/elephantine/pg"
+)
+
+func (s *PGDocStore) RunCleaner(ctx context.Context, period time.Duration) {
+	for {
+		select {
+		case <-time.After(period):
+		case <-ctx.Done():
+			return
+		}
+
+		jobLock, err := pg.NewJobLock(s.pool, s.logger, "cleaner", pg.JobLockOptions{
+			PingInterval:  10 * time.Second,
+			StaleAfter:    1 * time.Minute,
+			CheckInterval: 20 * time.Second,
+			Timeout:       5 * time.Second,
+		})
+		if err != nil {
+			s.logger.ErrorCtx(ctx, "failed to create job lock",
+				elephantine.LogKeyError, err)
+
+			continue
+		}
+
+		err = jobLock.RunWithContext(ctx, s.removeExpiredLocks)
+		if err != nil {
+			s.logger.ErrorCtx(
+				ctx, "lock cleaner error",
+				elephantine.LogKeyError, err,
+			)
+		}
+	}
+}
+
+func (s *PGDocStore) removeExpiredLocks(ctx context.Context) error {
+	s.logger.Debug("removing expired document locks")
+
+	err := s.reader.DeleteExpiredDocumentLocks(ctx,
```

Comment:
As per our earlier discussion in https://github.com/ttab/elephant-repository/pull/73 I would like all changes to document state to be done while holding a write lock on the document row. Even if it **should** be harmless to delete expired locks.

This could be done by fetching expired locks like this (untested queries):

``` sql
SELECT d.uuid, l.expires as lock_expires, l.app
FROM document_lock as l
     INNER JOIN document AS d ON d.uuid = l.uuid
WHERE l.expires < @cutoff
FOR UPDATE OF d SKIP LOCKED;
```

...and then in the same transaction:

``` sql
DELETE FROM document_lock
WHERE uuid = ANY(@uuids::text[]);
```

It becomes a two-phase thing instead of just a single query. But we'll know for certain that no concurrency issues will arise.